src/handler/mod.rs

```diff
@@ -85,7 +85,7 @@ pub enum HandlerIn {
     ///
     /// Note: To update an ENR for an unknown node, we request a FINDNODE with distance 0 to the
     /// `NodeContact` we know of.
-    Request(NodeContact, Box<Request>),
+    Request(RequestContact, Box<Request>),
```

Comment:
I could be missing something here, but I dont think this is needed.

For the topic advertisement stuff, I was thinking we didn't need to touch the handler at all. The handler should handle sessions,encryption and all that kind of magic for us. I thought we could just modify the service and send different RPC's through. The idea being that the service doesn't need to worry about establishing sessions etc thats all handled in the `handler`. The `Handler` sends and manages `Requests`. It doesn't need to know really whats inside the requests, so I dont think needs a `regtopic_with_ticket()` function. 

I did miss a bit of the discussion of the problem we were facing with obtaining ENRs. This is how I thought it would work: 

Service decides a bunch of nodes to send RegTopic to. When it finds nodes, it must get their ENR (from local routing table) in order to send a request. 
Once it has selected a few ENRs, it sends a `Request` to the handler with RegTopic. As it stands, whether there is a session or not, the handler will establish a session if it needs to, or use a current established one. From a service perspective, we dont need to worry, handler deals with all that magic. Once the response comes back from the service's `RegTopic` request, it will be either a `Ticket` response or a `RegConfirmation` response (we can specify what allowed types we have here: https://github.com/sigp/discv5/blob/master/src/rpc.rs#L205). The handler could also spit out an error, indicating we couldn't establish a session or contact the node. 

Given a Response from the handler, we can get the ENR of the node that responded because we have `active_requests` in the `Service` (https://github.com/sigp/discv5/blob/master/src/service.rs#L163). 
So when the `Service` handles a response from the handler, it can get the `NodeContact`. See it here: https://github.com/sigp/discv5/blob/master/src/service.rs#L586

And really we just have to add a match arm here: https://github.com/sigp/discv5/blob/master/src/service.rs#L612 to handle the `Ticket` or `RegConfirmation` responses. And inside this match we have access to the ENR of the node we are getting a response from. 

The overall idea is that the handler hides all the complexity of sessions/encryption and all that. All we should have to modify is `service.rs` to handle the high level RPC's. `rpc.rs` to add the different rpc's we can send and how to encrypt/decrypt them. 

Let me know if i'm missing something tho

Reply:
This part is already implemented: "Service decides a bunch of nodes to send RegTopic to. When it finds nodes, it must get their ENR (from local routing table) in order to send a request."

The issue has been getting the ENR later for re-sending REGTOPIC when we got a TICKET response from the first REGTOPIC, and the ticket wait time has expired. I see your solution: upon receiving the TICKET response, extract the ENR from the active request and save it along with the ticket. I didn't pay attention to the ```active_requests``` in Service.

Thanks Age!

Reply:
This I noticed however "it will be either a Ticket response or a RegConfirmation response" is contradicting the spec which, as I understood it states we should always issue a ticket. https://github.com/ethereum/devp2p/blob/master/discv5/discv5-wire.md#regtopic-request-0x07

Reply:
Yes. You're right. This is interesting then. The RegConfirmation still references a requestid. 
So I think similar to the `FindNode` request, the RegTopic can get multiple responses for the same request. 

The logic there we need to be careful about. I.e we only hold requests for so long before they expire and we dont accept any more responses. Also we decide how long to wait for a RegConfirmation. I suspect we just wait a single RPC request timeout amount.